hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -0,0 +1,403 @@
+        })
+
+        // TODO: In standing with other similar cases, should we be mutating elementType, or copyStreamable?
+        coerce[PNDArray](nd.pType2).elementType.setRequired(nd.pType2.required && allRequired)
+      }
+      case NDArraySlice(nd, slices) => {
+        nd.inferSetPType(env)
+        val childTyp = coerce[PNDArray](nd.pType2)
+
+        // TODO: do slices need nd env?
+        slices.inferSetPType(env)
+        val remainingDims = coerce[PTuple](slices.pType2).types.filter(_.isInstanceOf[PTuple])
+        PNDArray(childTyp.elementType, remainingDims.length)
+      }
+      case NDArrayMatMul(l, r) => {
+        l.inferSetPType(env)
+        r.inferSetPType(env)
+        val lTyp = coerce[PNDArray](l.pType2)
+        val rTyp = coerce[PNDArray](r.pType2)
+        PNDArray(lTyp.elementType, PNDArray.matMulNDims(lTyp.nDims, rTyp.nDims), lTyp.required && rTyp.required)
+      }
+      case NDArrayWrite(_, _) => PVoid
+      case MakeStruct(fields) => {
+        PStruct(fields.map {
+          case (name, a) => {
+            a.inferSetPType(env)
+            (name, a.pType2)
+          }
+        }: _*)
+      }
+      case SelectFields(old, fields) => {
+        old.inferSetPType(env)
+        val tbs = coerce[PStruct](old.pType2)
+        tbs.select(fields.toFastIndexedSeq)._1
+      }
+      case InsertFields(old, fields, fieldOrder) => {
+        old.inferSetPType(env)
+        val tbs = coerce[PStruct](old.pType2)
+
+        // TODO: does this need old environment? No name
+        val s = tbs.insertFields(fields.map(f =>  {
+          f._2.inferSetPType(env)
+          (f._1, f._2.pType2)
+        }))
+        fieldOrder.map { fds =>
+          assert(fds.length == s.size)
+          PStruct(fds.map(f => f -> s.fieldType(f)): _*)
+        }.getOrElse(s)
+      }
+      case GetField(o, name) => {
+        o.inferSetPType(env)
+        val t = coerce[PStruct](o.pType2)
+        if (t.index(name).isEmpty)
+          throw new RuntimeException(s"$name not in $t")
+        // TODO: Why is this not a copy?
+        val fd = t.field(name).typ
+        fd.setRequired(t.required && fd.required)
+      }
+      case MakeTuple(values) => {
+        PTuple(values.map(v => {
+          v.inferSetPType(env)
+          v.pType2
+        }).toFastIndexedSeq)
+      }
+      case GetTupleElement(o, idx) => {
+        o.inferSetPType(env)
+        val t = coerce[PTuple](o.pType2)
+        assert(idx >= 0 && idx < t.size)
+        val fd = t.types(idx)
+        fd.setRequired(t.required && fd.required)
+      }
+      case CollectDistributedArray(contexts, globals, contextsName, globalsName, body) => {
+        contexts.inferSetPType(env)
+        globals.inferSetPType(env)
+
+        body.inferSetPType(env.bind(contextsName -> contexts.pType2, globalsName -> globals.pType2))
+        PArray(body.pType2)
+      }
+      case ReadPartition(_, _, _, rowType) => PStream(PType.canonical(rowType))
+      case _ => PVoid
```

Comment:
this is causing your bugs

Reply:
In what way? I understand this is where the PVoid is generated (I added this), but it seems like the Array* IRs sometimes have Aggs and other IRs that we don't conduct type inference on. I am not certain how to handle those, since some of them (ApplySpecial seem to require algebraic types), and others (Agg) we've decided not to support.

Reply:
This may sound confrontational, sorry. Meant: how should I handle these strange nodes?

Reply:
It was not taken that way, no worries! 

I think the base case should be an exception (trying to infer the ptype of a node we don't support).

For now, that's probably not possible, and the base case should be `PType.canonical(node.type)`